x-pack/plugins/alerting/server/rules_client/lib/update_meta.ts

```diff
@@ -16,8 +16,10 @@ export function updateMeta<T extends Partial<RawRule>>(
   alertAttributes: T
 ): T {
   if (Object.hasOwn(alertAttributes, 'apiKey') || Object.hasOwn(alertAttributes, 'apiKeyOwner')) {
-    alertAttributes.meta = alertAttributes.meta ?? {};
-    alertAttributes.meta.versionApiKeyLastmodified = context.kibanaVersion;
+    const newAlertAttributes = JSON.parse(JSON.stringify(alertAttributes));
```

Comment:
Also, not shown in the PR context, there's a JS Doc comment saying ` * @deprecated Use updateMetaAttributes instead` - but there appear to be a lot of usages of this!

Can you find out whether we _should_ deprecate this and open an issue to do that later, or maybe we don't really want to treat it as deprecated?  Dunno.

Reply:
Its because we use a type that we generate from a schema.
RawRule type is read-only. That is why I added that `Mutable<>` type,. But although it looks recursive, it works only for two levels. 

OTOH, a function should not modify an object and return it back. IIRC We have only 2 cases that we need to clone an object, modify and return.

I am creating a separate PR for RawRule v3, I am already using the investigation issue as meta.
Probably we will be able to remove more types as we continue investigating.

Reply:
I am creating a separate PR for RawRule v3, I am already using the investigation issue as meta.
Probably we will be able to remove more types as we continue investigating.

Reply:
> a function should not modify an object and return it back.

I agree in principal.  However, the previous version of the code was doing exactly that.  And from the name, I'd say that it's doing what it says: updating something in-place (and happens to return it).

This code looks even more suspect to me now, sorry :-) :

- the `newAlertAttributes` is typed as `any`, and so it looks like (tried in vs code) you can add any random properties to this object and will get no warnings.
- the previous version of the code updated the value in-place, returning the input.  The new version will return either the input (as before), or an entirely new object.  So, things could get out of sync, if someone was expecting them to be the same.

Any chance we can do it the way it's done in x-pack/plugins/alerting/server/rules_client/lib/update_meta_attributes.ts ?  Not clear why there are two of these ...

Reply:
I think I can deprecate updateMeta, would be the best solution IMO.
Will create a separate PR for that too.

Reply:
It created more problems than I thought, i just changed it to use the same overwrite as in update_meta_attributes.

Reply:
Looks like besides the deprecation comment and function anme, `update_meta_attributes.ts` and `update_meta.ts` are exactly the same.  I'd think we could remove one of these ...

Reply:
Per discussion, I think we decided to open an issue to remove one of these later, since they are so similar.

Reply:
Created a follow-on issue to remove it: https://github.com/elastic/kibana/issues/197811